### PR TITLE
[SDAG] Turn umin into smin if the saturation pattern is broken

### LINKED ISSUE #85706: ARM USAT instruction not generated after #82478
Example: https://godbolt.org/z/rx1hzhhYz

https://github.com/llvm/llvm-project/pull/82478 canonicalizes signed minmax into unsigned which seems to be breaking some patterns.

### PR DESCRIPTION
**Pull request by dtcxzyw**
As we canonicalizes smin with non-negative operands into umin in the middle-end, the saturation pattern will be broken.
This patch reverts the transform in DAGCombine to fix the regression on ARM.

Fixes https://github.com/llvm/llvm-project/issues/85706.

### CODE REVIEW

**Review by arsenm — APPROVED** [review]

**Review by arsenm — APPROVED** [review]

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -5589,7 +5592,7 @@ SDValue DAGCombiner::visitIMINMAX(SDNode *N) {
     case ISD::UMAX: AltOpcode = ISD::SMAX; break;
     default: llvm_unreachable("Unknown MINMAX opcode");
     }
-    if (TLI.isOperationLegal(AltOpcode, VT))
+    if (IsSatBroken || TLI.isOperationLegal(AltOpcode, VT))
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5595]: Should we limit this to `(!LegalOperations && IsSatBroken) || TLI.isOperationLegal(AltOpcode, VT)`?

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5595]: What will happen after legalization on targets which don't use the SAT pattern and ISD::SMIN isn't a legal operation?

Comment (dtcxzyw) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:5595]: On risc-v without zbb:
```
define signext i32 @test_umin_smax_usat(i32 signext %x) {
entry:
  %v1 = tail call i32 @llvm.smax.i32(i32 %x, i32 0)
  %v2 = tail call i32 @llvm.umin.i32(i32 %v1, i32 255)
  ret i32 %v2
}
```
```
; Before
# %bb.0:                                # %entry
        sgtz    a1, a0
        neg     a1, a1
        and     a0, a1, a0
        li      a1, 255
        bltu    a0, a1, .LBB0_2
# %bb.1:                                # %entry
        li      a0, 255
.LBB0_2:                                # %entry
        ret
; After
# %bb.0:                                # %entry
        sgtz    a1, a0
        neg     a1, a1
        and     a0, a1, a0
        li      a1, 255
        blt     a0, a1, .LBB0_2
# %bb.1:                                # %entry
        li      a0, 255
.LBB0_2:                                # %entry
        ret
```

It just turns an illegal umin into an illegal smin.

**Comment by dtcxzyw** [comment]
@UsmanNadeem Any comments?

**Review by UsmanNadeem — APPROVED** [review]
Thanks!

**Pull request merged**